### ISSUE TITLE
False positive dead code warning

### ISSUE THREAD

**Issue by elichai** [rust-lang/rust#91784]
Given the following code: 

```rust
pub fn outer(a: u8) -> u8 {
    inner(a).unwrap()
}

#[derive(Debug)]
enum Error {
    A { c: u8 },
}

fn inner(c: u8) -> Result<u8, Error> {
    match c {
        7 => Ok(c),
        _ => Err(Error::A { c }),
    }
}
```

The current output is:

```
warning: field is never read: `c`
 --> src/main.rs:7:9
  |
7 |     A { c: u8 },
  |         ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

This warning is incorrect, as `c` is observed by the `unwrap` in `outer`

You can see here that it panics correctly with `c` but still complains that `c` is unused: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7c711a7318911453779c552e25f36edf

**Comment by ehuss** [comment]
Thanks for the report!  I'm going to close this as a duplicate of #88900, just to keep the discussion in one place.
